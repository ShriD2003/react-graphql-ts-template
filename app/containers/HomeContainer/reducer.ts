import produce from 'immer';
import { createActions } from 'reduxsauce';
import get from 'lodash/get';

export const { Types: homeContainerTypes, Creators: homeContainerCreators } = createActions({
  requestGetLaunchList: ['launchQuery'],
  successGetLaunchList: ['launchData'],
  failureGetLaunchList: ['launchListError'],
  clearLaunchList: {}
});
export const initialState = { launchQuery: null, launchData: {}, launchListError: null };

export const homeContainerReducer = (
  state = initialState,
  action: { type?: any; somePayload?: string; launchQuery?: any; launchData?: any; error?: any }
) =>
  produce(state, (draft) => {
    switch (action.type) {
      case homeContainerTypes.REQUEST_GET_LAUNCH_LIST:
        draft.launchQuery = action.launchQuery;
        break;
      case homeContainerTypes.CLEAR_LAUNCH_LIST:
        draft.launchQuery = null;
        draft.launchListError = null;
        draft.launchData = {};
        break;
      case homeContainerTypes.SUCCESS_GET_LAUNCH_LIST:
        draft.launchData = action.launchData.data;
        break;
      case homeContainerTypes.FAILURE_GET_LAUNCH_LIST:
        draft.launchListError = get(action.launchData.errors, 'message', 'something_went_wrong');
        break;
    }
  });

export default homeContainerReducer;
