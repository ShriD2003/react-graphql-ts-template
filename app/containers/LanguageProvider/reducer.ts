/*
 *
 * LanguageProvider reducer
 *
 */
import { createActions } from 'reduxsauce';
import produce from 'immer';
import { DEFAULT_LOCALE } from '@app/i18n';

type LanguageProviderTypes = {
  CHANGE_LOCALE: 'CHANGE_LOCALE';
};

type LanguageProviderAction = {
  type?: keyof LanguageProviderTypes;
  locale?: string;
};

type LanguageProviderCreators = {
  changeLocale: (locale: 'en') => LanguageProviderAction;
};

export const { Types: languageProviderTypes, Creators: languageProviderActions } = createActions<
  LanguageProviderTypes,
  LanguageProviderCreators
>({
  changeLocale: ['locale']
});

export const initialState = {
  locale: DEFAULT_LOCALE
};

/* eslint-disable default-case, no-param-reassign */
export const languageProviderReducer = (state = initialState, action: LanguageProviderAction) =>
  produce(state, (draft) => {
    switch (action.type) {
      case languageProviderTypes.CHANGE_LOCALE:
        if (action.locale) {
          draft.locale = action.locale;
        }
        break;
    }
  });

export default languageProviderReducer;
