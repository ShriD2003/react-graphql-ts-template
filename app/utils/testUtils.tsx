import React from 'react';
import { I18nProvider } from '@lingui/react';
import { i18n } from '@lingui/core';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { Route, Router } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';
import configureStore from '@app/configureStore';
import { DEFAULT_LOCALE, translationMessages } from '@app/i18n';
import ConnectedLanguageProvider from '@containers/LanguageProvider';
import history from './history';

export const renderWithIntl = (children: React.ReactNode) => {
  i18n.load(DEFAULT_LOCALE, translationMessages[DEFAULT_LOCALE]);
  i18n.activate(DEFAULT_LOCALE);

  return render(<I18nProvider i18n={i18n}> {children} </I18nProvider>);
};

export const getComponentStyles = (Component: React.FC<any>, props = {}) => {
  renderWithIntl(Component(props));
  const { styledComponentId } = Component(props)!.type;
  const componentRoots = document.getElementsByClassName(styledComponentId);
  return window.getComputedStyle(componentRoots[0]);
};

export const renderProvider = (children: React.ReactNode, { path }: { path?: string } = {}, renderFn = render) => {
  const store = configureStore({}).store;
  return renderFn(
    <Provider store={store}>
      <ConnectedLanguageProvider messages={translationMessages}>
        <ThemeProvider
          theme={{
            main: 'violet'
          }}
        >
          <Router history={history}>{path ? <Route path={path}>{children}</Route> : children}</Router>
        </ThemeProvider>
      </ConnectedLanguageProvider>
    </Provider>
  );
};
export const timeout = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));
export const apiResponseGenerator = <Data,>(ok: boolean, data: Data, error?: object) => ({
  ok,
  data,
  error
});
