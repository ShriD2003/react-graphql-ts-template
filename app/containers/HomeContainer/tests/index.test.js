import React from 'react';
import { timeout, renderProvider } from '@utils/testUtils';
import { HomeContainerTest as HomeContainer, mapDispatchToProps } from '../index';
import { homeContainerTypes } from '../reducer';

describe('<HomeContainer /> tests', () => {
  let submitSpy;

  beforeEach(() => {
    submitSpy = jest.fn();
  });
  it('should render and match the snapshot', () => {
    const { baseElement } = renderProvider(
      <HomeContainer dispatchLaunchList={submitSpy} dispatchClearLaunchList={jest.fn()} />
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('should call dispatchClearLaunchList on page reload', async () => {
    const getLaunchListSpy = jest.fn();
    const clearLaunchListSpy = jest.fn();
    renderProvider(
      <HomeContainer dispatchClearLaunchList={clearLaunchListSpy} dispatchLaunchList={getLaunchListSpy} />
    );
    await timeout(500);
    expect(clearLaunchListSpy).toBeCalled();
  });

  it('should call dispatchLaunchList on page reload', async () => {
    renderProvider(<HomeContainer dispatchLaunchList={submitSpy} dispatchClearLaunchList={jest.fn()} />);
    await timeout(500);
    expect(submitSpy).toBeCalledWith();
  });

  it('should validate mapDispatchToProps actions', async () => {
    const dispatchLaunchListSpy = jest.fn();
    const actions = {
      dispatchLaunchList: { type: homeContainerTypes.REQUEST_GET_LAUNCH_LIST },
      dispatchClearLaunchList: { type: homeContainerTypes.CLEAR_LAUNCH_LIST }
    };

    const props = mapDispatchToProps(dispatchLaunchListSpy);
    props.dispatchLaunchList();
    expect(dispatchLaunchListSpy).toHaveBeenCalledWith(actions.dispatchLaunchList);

    await timeout(500);
    props.dispatchClearLaunchList();
    expect(dispatchLaunchListSpy).toHaveBeenCalledWith(actions.dispatchClearLaunchList);
  });

  it('should render default error message when search goes wrong', () => {
    const defaultError = 'This is an error';
    const { getByTestId } = renderProvider(
      <HomeContainer
        launchListError={defaultError}
        dispatchClearLaunchList={jest.fn()}
        dispatchLaunchList={submitSpy}
      />
    );
    expect(getByTestId('error-message')).toBeInTheDocument();
    expect(getByTestId('error-card')).toBeInTheDocument();
    expect(getByTestId('error-message').textContent).toBe(defaultError);
  });

  it('should render the data when loading becomes false', () => {
    const launchData = {
      launches: [
        {
          missionName: 'Sample Mission',
          launchTimeLocal: 'sample time',
          links: {
            wikipedia: 'sample link'
          }
        }
      ]
    };
    const { getByTestId } = renderProvider(
      <HomeContainer launchData={launchData} dispatchLaunchList={submitSpy} dispatchClearLaunchList={jest.fn()} />
    );
    expect(getByTestId('list')).toBeInTheDocument();
  });

  it('should render Skeleton Comp when "loading" is true', async () => {
    const { baseElement } = renderProvider(
      <HomeContainer dispatchLaunchList={submitSpy} dispatchClearLaunchList={jest.fn()} loading={true} />
    );

    expect(baseElement.getElementsByClassName('ant-skeleton').length).toBe(1);
  });
});
