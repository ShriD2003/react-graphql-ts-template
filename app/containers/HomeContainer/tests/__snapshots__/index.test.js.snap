// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomeContainer /> container tests should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="HomeContainer__Container-wd79o1-1 ACgFJ"
    >
      <div
        class="ant-card ant-card-bordered HomeContainer__CustomCard-wd79o1-0 fZjvjB"
      >
        <div
          class="ant-card-head"
        >
          <div
            class="ant-card-head-wrapper"
          >
            <div
              class="ant-card-head-title"
            >
              Spacex Launches Search
            </div>
          </div>
        </div>
        <div
          class="ant-card-body"
        />
      </div>
    </div>
  </div>
</body>
`;
